example/fine_advection/advection_package.cpp

```diff
@@ -94,12 +94,59 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
   pkg->AddField<Conserved::divD>(
       Metadata({Metadata::Cell, Metadata::Derived, Metadata::OneCopy}));
 
-  pkg->CheckRefinementBlock = CheckRefinement;
+  bool check_refine_mesh =
+      pin->GetOrAddBoolean("parthenon/mesh", "CheckRefineMesh", true);
+  if (check_refine_mesh) {
+    pkg->CheckRefinementMesh = CheckRefinementMesh;
+  } else {
+    pkg->CheckRefinementBlock = CheckRefinement;
+  }
   pkg->EstimateTimestepMesh = EstimateTimestep;
   pkg->FillDerivedMesh = FillDerived;
   return pkg;
 }
 
+void CheckRefinementMesh(MeshData<Real> *md,
+                         parthenon::ParArray1D<AmrTag> &delta_levels) {
+  // refine on advected, for example.  could also be a derived quantity
+  static auto desc = parthenon::MakePackDescriptor<Conserved::phi>(md);
+  auto pack = desc.GetPack(md);
+
+  auto pkg = md->GetMeshPointer()->packages.Get("advection_package");
+  const auto &refine_tol = pkg->Param<Real>("refine_tol");
+  const auto &derefine_tol = pkg->Param<Real>("derefine_tol");
+
+  auto ib = md->GetBoundsI(IndexDomain::entire);
+  auto jb = md->GetBoundsJ(IndexDomain::entire);
+  auto kb = md->GetBoundsK(IndexDomain::entire);
+  auto scatter_levels = delta_levels.ToScatterView<Kokkos::Experimental::ScatterMax>();
+  parthenon::par_for_outer(
+      PARTHENON_AUTO_LABEL, 0, 0, 0, pack.GetNBlocks() - 1, 0,
+      pack.GetMaxNumberOfVars() - 1, kb.s, kb.e,
+      KOKKOS_LAMBDA(parthenon::team_mbr_t team_member, const int b, const int n,
+                    const int k) {
+        typename Kokkos::MinMax<Real>::value_type minmax;
+        par_reduce_inner(
+            parthenon::inner_loop_pattern_ttr_tag, team_member, jb.s, jb.e, ib.s, ib.e,
+            [&](const int j, const int i,
+                typename Kokkos::MinMax<Real>::value_type &lminmax) {
+              lminmax.min_val = (pack(n, k, j, i) < lminmax.min_val ? pack(n, k, j, i)
+                                                                    : lminmax.min_val);
+              lminmax.max_val = (pack(n, k, j, i) > lminmax.max_val ? pack(n, k, j, i)
+                                                                    : lminmax.max_val);
+            },
+            Kokkos::MinMax<Real>(minmax));
+
+        auto levels_access = scatter_levels.access();
+        auto flag = AmrTag::same;
+        if (minmax.max_val > refine_tol && minmax.min_val < derefine_tol)
+          flag = AmrTag::refine;
+        if (minmax.max_val < derefine_tol) flag = AmrTag::derefine;
+        levels_access(b).update(flag);
```

Comment:
Does this work as expected (in particular with the split across `k`, and `j/i` (similarly below for the derivs)?
If we refine just a single cell in a block, this info should be persistent even if other cells in that block is `same` or `derefine`, so currently there could be a clash of info across `k`s doesn't it?

Reply:
I think so. It should be equivalent to doing
```cpp
Kokkos::atomic_max(&delta_levels(b), flag);
```
So the race condition is across `k`s and the `max` ensures that refinement always wins

Reply:
Ah, right, yes. I didn't see see. Might be worth adding a comment in that direction (at least to the common/non-example function in `src/amr_criteria/refinement_package.cpp`.